File: client/java-client/src/main/java/com/kotak/neo/client/exceptions/ApiException.java
package com.kotak.neo.client.exceptions;

public class ApiException extends OpenApiException {
    public final int status;
    public final String reason;
    public final String body;

    public ApiException(int status, String reason, String body) {
        super(String.format("(%d) %s%s", status, reason, body == null ? "" : ": " + body));
        this.status = status;
        this.reason = reason;
        this.body = body;
    }
}

File: client/dotnet-client/NeoApi/Exceptions/ApiException.cs
namespace Kotak.Neo.Exceptions;

public class OpenApiException : Exception
{
    public OpenApiException(string message) : base(message) { }
}

public class ApiTypeError : OpenApiException
{
    public ApiTypeError(string msg) : base(msg) { }
}

public class ApiValueError : OpenApiException
{
    public ApiValueError(string msg) : base(msg) { }
}

public class ApiAttributeError : OpenApiException
{
    public ApiAttributeError(string msg) : base(msg) { }
}

public class ApiKeyError : OpenApiException
{
    public ApiKeyError(string msg) : base(msg) { }
}

public class ApiException : OpenApiException
{
    public int Status { get; }
    public string Reason { get; }
    public string? Body { get; }

    public ApiException(int status, string reason, string? body = null)
        : base($"({status}) {reason}{(body is null ? "" : $": {body}")}")
    {
        Status = status;
        Reason = reason;
        Body = body;
    }
}
